addons/hw_drivers/websocket_client.py

```diff
@@ -13,6 +18,68 @@
+WS_REQUEST_TIME_TIMEOUT = 120  # seconds
+
+
+def _check_ws_request_pool():
+    """Check if the request pool is not stuck with timeout-ing requests.
+    We can't kill a thread, but if the situation is desperate we will restart the IoT.
+
+    We use a concurrent.futures.ThreadPoolExecutor to handle the websocket messages.
+    Ccenario example:
+     When a ws request R1 is received, it will create a thread T1 in the pool to handle the request
+     When a second ws request is received R2:
+      if R1 finished, T1 is "free" so it will be reused
+      if R1 is still running, T1 will be busy so a second thread T2 will be created to handle R2
+     ...
+     until the maximum number of threads is reached (see `ThreadPoolExecutor._max_workers`).
+     In this case they will wait until a thread is free.
+    """
+    def thread_stack_trace(thread):
+        frame = sys._current_frames().get(thread.ident, None)
+        if frame:
+            return ". Thread stack:\n" + ''.join(traceback.format_stack(frame))
+        return ''
+
+    delay_default = min(WS_REQUEST_TIME_WARN * 2, WS_REQUEST_TIME_TIMEOUT)
+    check_delay = delay_default
+    while True:
+        time.sleep(check_delay)
+
+        _logger.debug("Checking websocket request pool with current threads: %s", WS_REQUEST_POOL._threads)
+        check_delay = delay_default
+        stuck_threads = []
+        for ws_thread in WS_REQUEST_POOL._threads:
+            if not hasattr(ws_thread, 'starting_time'):
+                # thread that was started by a previous ws request but currently idle
+                continue
+
+            time_differece = time.time() - ws_thread.starting_time
+            _logger.debug("Websocket thread %s is busy since %.2f sec", ws_thread, time_differece)
+            if time_differece > WS_REQUEST_TIME_TIMEOUT:
+                _logger.error("Websocket thread %s is stuck (busy since %.2f sec)%s", ws_thread, time_differece, thread_stack_trace(ws_thread))
+                stuck_threads.append(ws_thread)
+                check_delay = min(WS_REQUEST_TIME_TIMEOUT // 3, check_delay)  # Check more often if there are stuck threads
+
+            elif time_differece > WS_REQUEST_TIME_WARN:
+                _logger.warning("Websocket thread %s may be stuck (busy since %.2f sec)%s", ws_thread, time_differece, thread_stack_trace(ws_thread))
+                check_delay = min(WS_REQUEST_TIME_TIMEOUT // 2, check_delay)  # Check a bit more often if there are no stuck threads
+
+        if len(stuck_threads) >= WS_REQUEST_POOL._max_workers // 2:
+            error_message = 'Too many websocket threads are stuck. Restarting the IoT.'
+            for stuck_thread in stuck_threads:
+                error_message += '\nThread %s (started at: %s) timeout%s' % (stuck_thread, stuck_thread.starting_time, thread_stack_trace(stuck_thread))
+            _logger.critical(error_message)
+            WS_REQUEST_POOL.shutdown(wait=False)
+            helpers.odoo_restart(2)
```

Comment:
restarting iot is too violent and unneeded.
In PrinterDriver_W create a system that will manage the blocked ghostscript.Ghostscript class instance and kill it after a timeout